sys/analog_util/adc_util.c

```diff
@@ -14,28 +15,52 @@
  * @brief       ADC utility function implementation
  *
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Joakim Nohlgård <joakim.nohlgard@eistec.se>
  *
  * @}
  */
 
+#include <stdint.h>
+#include <inttypes.h>
+#include "cpu.h"
+#include "periph/adc.h"
 #include "analog_util.h"
 
-/* keep a max value to ADC resolution mapping for quick access in the ROM */
-static const int val_max[] = {
-    [ADC_RES_6BIT] =    0x003f,
-    [ADC_RES_8BIT] =    0x00ff,
-    [ADC_RES_10BIT] =   0x03ff,
-    [ADC_RES_12BIT] =   0x0fff,
-    [ADC_RES_14BIT] =   0x3fff,
-    [ADC_RES_16BIT] =   0xffff
-};
+#define ENABLE_DEBUG 0
+#include "debug.h"
 
-int adc_util_map(int sample, adc_res_t res, int min, int max)
+/**
+ * @brief Convert adc_res_t resolution setting into numeric bit count
+ */
+static unsigned int _adc_res_bits(adc_res_t res)
+{
+    switch (res) {
+        case ADC_RES_6BIT:
+            return 6;
+        case ADC_RES_8BIT:
+            return 8;
+        case ADC_RES_10BIT:
+            return 10;
+        case ADC_RES_12BIT:
+            return 12;
+        case ADC_RES_14BIT:
+            return 14;
+        case ADC_RES_16BIT:
+        default:
+            return 16;
+    }
+}
+
+int32_t adc_util_map(int sample, adc_res_t res, int32_t min, int32_t max)
 {
-    return ((((max - min) * sample) / val_max[res]) + min);
+    /* Using 64 bit signed int as intermediate to prevent overflow when range
+     * multiplied by sample requires more bits than int offers */
+    int32_t scaled = (((int64_t)(max - min) * sample) >> _adc_res_bits(res));
```

Comment:
while it makes sense to use `int64_t` for(correct) internal maths, I don't understand why you changed parameters and return type to `int32_t` when our ADC API returns only `int`

Reply:
I changed the return type to int32_t because the scaled value may very well be useful to express larger numbers than what int offers on the smaller architectures. Using int32_t means we can use min and max values > 32768 safely on all architectures. There are many situations where the values are more useful if you can convert them to some scientific measure directly, and using a wider data type mean we can perform the conversion using integer math without needing to shift down and losing more precision. This function is intended to be usable in more than just for creating phydat_t values.